File: src/cpp/128.cpp
#include <unordered_set>
#include <vector>

class Solution {
public:
  int longestConsecutive(std::vector<int> &nums) {
    // the problem is to find the longest consecutive sequence in an array of
    // numbers.
    //
    // the sequence does not have to be contiguous.
    // **the sequence does not have to appear in order like the original array**
    // elements in the sequence must be +1 larger than the previous.
    int longest = 0;

    // convert to a set so we have fast lookup times
    std::unordered_set<int> numSet(nums.begin(), nums.end());

    for (int n : numSet) {
      // find the start of a sequence
      if (numSet.contains(n - 1)) {
        continue;
      }

      // scan through the set of numbers to find the next and next and next
      int running = 1;
      while (numSet.contains(n + running)) {
        running++;
      }
      longest = std::max(longest, running);
    }

    return longest;
  }
};

File: src/python/128.py
from typing import List


class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        # build the set of all numbers first (drops duplicates and allows for set member set check)
        seen = set(nums)

        longest = 0
        # step through each number (seen not nums)
        for num in seen:
            # if the number before this number is not in the seen set, then this is the start of a sequence
            if num - 1 not in seen:
                length = 1
                # we're at the head of a sequence, count until we can't anymore
                while num + 1 in seen:
                    num += 1
                    length += 1
                longest = max(longest, length)

        return longest
